### PR TITLE
Update Adapter LDAP FATs to use UnboundID instead of EmbeddedApacheDS

### PR DESCRIPTION
**Pull request by ryanesch**

### CODE REVIEW

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
#build

**Comment by ryanesch** [comment]
#build

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPReferralTest.java [hunk 1]
```diff
@@ -178,35 +182,47 @@ private static void setupLibertyServer() throws Exception {
...
+                                             InMemoryListenerConfig.createLDAPConfig("LDAP", // Listener name
+                                                                                     null, // Listen address. (null = listen on all interfaces)
+                                                                                     0, // Listen port (0 = automatically choose an available port)
+                                                                                     null) // StartTLS factory
+        ); // Client factory
+        subordinateConfig.setSchema(null);
+        dsSubordinate = new InMemoryDirectoryServer(subordinateConfig);
+        dsSubordinate.startListening();
+
+//        Entry entry = new Entry(DELEGATE_DN);
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPReferralTest.java]: Delete this commented out code.

File: dev/com.ibm.ws.com.unboundid/resources/wimschema.ldif [hunk 2]
```diff
@@ -0,0 +1,225 @@
+dn: cn=schema
```
Comment (jvanhill) [inline dev/com.ibm.ws.com.unboundid/resources/wimschema.ldif]: Add copyright. Use '#' for comment lines.

**Comment by ryanesch** [comment]
#build

File: dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    private InMemoryDirectoryServer ds = null;
+
+    /** The name for this server instance. */
+    private final String name = null;
+
+    /**
+     * Creates a new instance of the in memory LDAP server. It initializes the directory
+     * service.
+     *
+     * @param instance The name of this ADS instance.
```
Comment (jvanhill) [inline dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java]:
```suggestion
     * @param bases The base entries to create for this in-memory LDAP server.
```

File: dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+/**
+ * An in memory UnboundID LDAP server. See ContextPoolTimeoutTest for an example on how to use this.
+ * Basic steps: Create an instance of this class passing in your base DNs, eg. o=ibm,c=us. Add users
+ * and groups via entries. Note you will need to build the LDAP tree, so if you want to create
+ * uid=user1,ou=users,o=ibm,c=us, then you need an entry for ou=users first. The WIM schema has
+ * been merged in, so WIM attributes are available when using wimInetOrgPerson object class.
+ * To simulate Active Directory, use simulatedMicrosoftSecurityPrincipal object class. This
+ * allows memberOf and samAccountName to be used. You can view the WIM schema in wimschema.ldif.
+ */
```
Comment (jvanhill) [inline dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java]: I would maybe re-format this, though the content is good. You might want to mention the wimGroupOfNames class as well. Maybe a list of objectclasses a user would use with a short description of what they are?

File: dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        config.addAdditionalBindCredentials(getBindDN(), getBindPassword());
+        config.setListenerConfigs(
+                                  InMemoryListenerConfig.createLDAPConfig("LDAP", // Listener name
+                                                                          null, // Listen address. (null = listen on all interfaces)
+                                                                          0, // Listen port (0 = automatically choose an available port)
+                                                                          null) // StartTLS factory
+        ); // Client factory
+
+        /*
+         * Merge the default schema with our WIM schema. The WIM schema adds wimInetOrgPerson objectclass which
```
Comment (jvanhill) [inline dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java]: wimGroupOfNames, simulatedMic.... etc

File: dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+ */
+public class InMemoryLDAPServer {
+
+    private static final Class<?> c = InMemoryLDAPServer.class;
+
+    private InMemoryDirectoryServerConfig config = null;
+    private InMemoryDirectoryServer ds = null;
+
+    /** The name for this server instance. */
+    private final String name = null;
```
Comment (jvanhill) [inline dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java]: We don't seem to use this.
```suggestion

```

File: dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    public static String getBindPassword() {
+        return "secret";
+    }
+
+    /**
+     * Get the name of this LDAP server instance.
+     *
+     * @return The name.
+     */
+    public String getInstanceName() {
```
Comment (jvanhill) [inline dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java]: Delete this method.

File: dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    public void add(Entry entry) throws LDAPException {
+        ds.add(entry);
+    }
+
+    /**
+     * Stop the InMemoryDirectoryServer.
+     */
+    public void stopServer() {
```
Comment (jvanhill) [inline dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java]:
```suggestion
    public void shutDown() {
```

File: dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    public InMemoryDirectoryServer getLdapServer() {
+        return ds;
+    }
+
+    public void add(Entry entry) throws LDAPException {
+        ds.add(entry);
+    }
+
+    /**
+     * Stop the InMemoryDirectoryServer.
```
Comment (jvanhill) [inline dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java]:
```suggestion
     * Shut down the InMemoryDirectoryServer. This is a convenience method to call shutDown(true).
```

File: dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+     */
+    public void stopServer() {
+        ds.shutDown(true);
+    }
+
+    /**
```
Comment (jvanhill) [inline dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java]: Shut down the InMemoryDirectoryServer.
```suggestion
    /**
     * Shut down the InMemoryDirectoryServer.
     *
```

File: dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+     * @param b true to close all existing connections, or false to stop accepting new connections.
+     */
+    public void shutDown(boolean b) {
+        ds.shutDown(b);
+
+    }
+
+    /**
```
Comment (jvanhill) [inline dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java]: Method description.

File: dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+     * @return the port this directory server is listening to
+     */
+    public int getListenPort() {
+        return ds.getListenPort();
+    }
+
+    /**
```
Comment (jvanhill) [inline dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java]: Method description.

File: dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+     * @param userDn
+     * @param modification
+     * @throws LDAPException
+     */
+    public void modify(String userDn, Modification modification) throws LDAPException {
+        ds.modify(userDn, modification);
+
+    }
+
+    /**
```
Comment (jvanhill) [inline dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java]: Method description.

File: dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+     * @param dn the DN to remove from the directory server
+     * @throws LDAPException
+     */
+    public void delete(String dn) throws LDAPException {
+        DeleteRequest deleteRequest = new DeleteRequest(dn);
+        delete(deleteRequest);
+    }
+
+    /**
```
Comment (jvanhill) [inline dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java]: Method description.

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPReferralTest.java [hunk 15]
```diff
@@ -80,12 +80,12 @@
      */
     private static ServerConfiguration emptyConfiguration = null;
 
-    private static EmbeddedApacheDS delegateServer = null;
     private static final String DELEGATE_DN = "o=ibm,c=us";
 //    private static final String DELEGATE_USER_DN = "uid=user1," + DELEGATE_DN;
 //    private static final String DELEGATE_GROUP_DN = "cn=group1," + DELEGATE_DN;
+    static InMemoryLDAPServer dsSubordinate;
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPReferralTest.java]:
```suggestion
    private static InMemoryLDAPServer dsSubordinate;
```

File: dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPReferralTest.java [hunk 16]
```diff
@@ -80,12 +80,12 @@
...
+    static InMemoryLDAPServer dsDelegate;
```
Comment (jvanhill) [inline dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/LDAPReferralTest.java]:
```suggestion
    private static InMemoryLDAPServer dsDelegate;
```

File: dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+/**
+ * An in memory UnboundID LDAP server. See ContextPoolTimeoutTest for an example on how to use this.
+ *
+ * <br><br>
+ * Basic steps: Create an instance of this class passing in your base DNs, eg. o=ibm,c=us. Add users
+ * and groups via entries. Note you will need to build the LDAP tree, so if you want to create
+ * uid=user1,ou=users,o=ibm,c=us, then you need an entry for ou=users first.
+ *
+ * <br><br>
+ * The schema supports the following classes:
```
Comment (jvanhill) [inline dev/com.ibm.ws.com.unboundid/src/com/ibm/ws/com/unboundid/InMemoryLDAPServer.java]:
```suggestion
 * In addition to the standard classes, the schema has been extended with the following classes:
```

**Pull request merged**